### ISSUE TITLE
"eXist-db" app remains running after first install on macOS

### ISSUE THREAD

**Issue by joewiz** [eXist-db/exist#1248]
When I first run eXist-db.app using current nightly builds on macOS, the eXist-db icon that appears in my dock doesn't close when I quit eXist via the menu bar.  It seems the only way to get this icon to go away is to control-click on it and select quit.  

This is odd, since it's not associated with the eXist-db process. It seems harmless but will certainly confuse users. 

The "normal" behavior I would expect from eXist-db is that the dock icon appears briefly and then disappears after the menu bar icon appears.  

Since this persistent dock icon only appears once - on first run of a new installation - my hunch is that this is somehow related to something that only happens on first run and not subsequent runs.  But I'm not sure what it is.

**Comment by duncdrum** [comment]
From the activity monitor screenshot in #1275 it appears to be doing something….

**Comment by adamretter** [comment]
@dizzzz I don't think I am the correct person for this; I always wanted to drop the IzPack installer and Mac DMG etc ;-)

**Comment by joewiz** [comment]
(I've added a proposed topic for a future call about dropping the IzPack installer and the Mac DMG. See the current draft of the [2017-02-27 - eXist Community Call](https://docs.google.com/document/d/116O2kVM5_7QNsL7DVe21y-GFM7OQKi1FjfD8NLnqFXQ/edit?usp=sharing) under Agenda section 5.a. Any topic is ok to propose, but I'm especially trying to capture knotty issues like this for future discussion.)

**Comment by adamretter** [comment]
@joewiz Awesome, I think I would prefer to see platform native packages, but on Mac maybe use your HomeBrew, but have it build from a .tar.gz.

**Comment by duncdrum** [comment]
Just a comment on the homebrew vs  .jar vs .dmg topic for mac OS. How the app gets into the applications folder is less of an issue then the different default data directories of each installer. This could make for some very unpleasant upgrading with either an old directory collecting dust in a hidden systems folder, or with accidental overwrites which could break restorability. Especially going from 2.2 to 3.x.

**Comment by duncdrum** [comment]
same with 3.1 installer

**Comment by joewiz** [comment]
I just noticed several things that might explain why first-runs of 3.0 (release) and now also 3.1.0 result in the eXist-db app icon remaining active even after the app has completed launching the eXist-db process, whereas this was not the case with 2.2 or 3.0RC1. The differences are evident when comparing 3.0RC1 to 3.1.0, while monitoring the dock, menu bar, and Activity Monitor during first launch. 

In 3.0RC1, here is the first-launch sequence:

1. The app's dock icon bounces, and an `eXist-db` process appears in Activity Monitor.
2. After a few bounces the app icon disappears from the Dock, and first process disappears from Activity Monitor.
3. A second `eXist-db` process appears in Activity Monitor, the eXist-db Configuration window appears, and the eXist-db menu bar icon appears.
4. Upon dismissing the eXist-db Configuration window (by selecting the Save button and approving the creation of the data directory in `~/Library/Application Support/org.exist`), the second process disappears.
5. A third `eXist-db` process appears in Activity Monitor, and the eXist-db splash screen appears. This third `eXist-db` process remains active in Activity Monitor until the database is quit via the eXist-db menu bar > `Quit (and stop server)`. 

In 3.1.0, the sequence is remarkably different:

1. The app's dock icon bounces, and an `eXist-db` process appears in Activity Monitor.
2. After a few bounces the app's dock icon stops bouncing, but remains active in the dock. The process in the Activity Monitor disappears, as the eXist-db Configuration window appears. While this window is open, no eXist-db processes remain in the Activity Monitor!
3. Upon dismissing the eXist-db Configuration window (by selecting the Save button and approving the creation of the data directory in `~/Library/Application Support/org.exist`), a second process appears in Activity Monitor, now called `"eXist-db"` (note the double quotes), the eXist-db menu bar icon appears, and the eXist-db splash screen appears. 
4. The `"eXist-db"` process remains active in the Activity Monitor until the database is quit. The eXist-db app icon in the dock remains active **even after** the database is quit via the eXist-db menu bar > `Quit (and stop server)`. 

(One other small difference: Besides the data directory in `~/Library/Application Support/org.exist`, eXist-db 3.1.0 also creates a folder in `~/Library/Saved Application State/org.exist.start.Main.savedState`, which 3.0RC1 did not create.)

Key differences here:

* In 3.0RC1, the eXist processes are only ever called `eXist-db`. In 3.1.0, the eXist processes appear first as `eXist-db` but then `"eXist-db"`
* The sequences of processes is quite different - 3 processes in 3.0RC1 and 2 processes in 3.1.0.
* In 3.0RC1 the app's dock icon disappears after a few bounces, while in 3.1.0 it remains active until manually quit.

To reproduce these "first run" sequences, you need to:

1. Delete `/Applications/eXist-db.app`
2. Delete `~/Library/Application Support/org.exist` (and if present, `~/Library/Saved Application State/org.exist.start.Main.savedState`)
3. Drag `eXist-db.app` from the mounted disk image into the `/Applications` folder
4. Start eXist, while monitoring the dock, menu bar, and Activity Monitor
5. Repeat to test again.

Again, I'm using macOS 10.12.3 (16D32) with Java 1.8.0_121-b13.

Does this provide any clues about what could be responsible for keeping the app's dock icon active in 3.1.0?

**Comment by shabanovd** [comment]
@joewiz can you try to get threads dump of that ghost process? (my guess, that is jvm process).

**Comment by joewiz** [comment]
@shabanovd Thanks, I'd be happy to try. I have some old notes on generating a thread dump - below. Is this what I should do? Or is there a different method you have in mind? And is there a specific one of the 2-3 processes detailed above that you'd like to see, or as many as possible?

```
- get java process id #:

$ ps aux | grep java

- the result will look something like 

root     20232  0.3  0.0 114440  1376 ?        Sl   20:29   0:17 /data/paho-svn/trunk/exist/tools/wrapper/bin/./wrapper-linux-x86-64 /data/paho-svn/trunk/exist/tools/wrapper/bin/../conf/wrapper.conf wrapper.syslog.ident=eXist-db wrapper.pidfile=/data/paho-svn/trunk/exist/tools/wrapper/bin/./eXist-db.pid wrapper.name=eXist-db wrapper.displayname=eXist-db Native XML Database wrapper.daemonize=TRUE wrapper.statusfile=/data/paho-svn/trunk/exist/tools/wrapper/bin/./eXist-db.status wrapper.java.statusfile=/data/paho-svn/trunk/exist/tools/wrapper/bin/./eXist-db.java.status wrapper.lockfile=/var/lock/subsys/eXist-db wrapper.script.version=3.5.17
root     20234  104 44.2 6903744 3386852 ?     Sl   20:29  99:49 /usr/java/jdk1.6.0_26/bin/java -Djava.endorsed.dirs=../../../lib/endorsed -Dfile.encoding=UTF-8 -Dexist.home=../../.. -Djetty.home=../../../tools/jetty -Dlog4j.configuration=file:../conf/log4j.xml -XX:-HeapDumpOnOutOfMemoryError -Xms256m -Xmx5000m -Djava.library.path=../lib -classpath ../classes:../lib/wrapper.jar:../../../start.jar -Dwrapper.key=9A_WI1X0CWnyjj0R -Dwrapper.port=32000 -Dwrapper.jvm.port.min=31000 -Dwrapper.jvm.port.max=31999 -Dwrapper.disable_console_input=TRUE -Dwrapper.pid=20232 -Dwrapper.version=3.5.17 -Dwrapper.native_library=wrapper -Dwrapper.arch=x86 -Dwrapper.service=TRUE -Dwrapper.cpu.timeout=10 -Dwrapper.jvmid=1 org.exist.wrapper.Main jetty ../../../tools/jetty/etc/jetty.xml
ec2-user 20672  0.0  0.0 103144   800 pts/0    S+   22:05   0:00 grep java

- make note of the # after 'root' on the 1st line - 20232 in this example

- trigger thread dump in log, using the process ID #:

$ sudo kill -SIGHUP 20232
$ sudo kill -SIGQUIT 20232

- Shut down eXist
```

**Comment by shabanovd** [comment]
@joewiz yes, you should call `kill -3 20234` (java process) and check logs for dump.

**Comment by adamretter** [comment]
@joewiz If it is easier you can just call `jstack 20234` and that will give you the stacktrace directly in the console

**Comment by joewiz** [comment]
Superseded by #1431

**Comment by joewiz** [comment]
Mistakenly thought this was the same as #1431. That issue is different and seems more serious.

**Comment by joewiz** [comment]
After further study, this issue is not specific to the eXist-db.app application, but can be reproduced on macOS even when building eXist from source. Steps to reproduce:

1. Build eXist
2. Delete the `vm.properties` file to cause eXist's launcher to think eXist is starting up for the first time: `rm vm.properties`
3. Start eXist via `open start.jar` or double-clicking on start.jar. 
4. A new icon to appear in the dock, and the "eXist-db configuration" dialog will appear. Select the "Save" button, and "OK" on the data directory dialog.
5. The eXist splash screen and task bar icon will appear, but notice that the dock icon remains.
6. From the task bar icon, select "Quit". The task bar icon will disappear, but notice the dock icon remains.
7. The task bar icon must be manually quit, otherwise this process remains: 

> $ ps aux | grep java
> joe              17980   0.0  1.3 11162816 215140   ??  S     4:43PM   0:05.47 /usr/bin/java -jar /Users/joe/workspace/exist/start.jar

As @adamretter suggested above, I've taken a jstack of the ghost process, which I've pasted in below.

Finally, if it's any help in tracking down the cause, I am pretty sure that this bug appeared with this commit by @wolfgangmm on Jul 29, 2016, which was designed to address bugs in the macOS launcher: https://github.com/eXist-db/exist/commit/f37b6fd5953aa02a634c9f2a751cf75d7144c200. The bug was not present in eXist 3.0-RC1 (released Jul 7, 2015) but was present in 3.0-RC2 (Aug 3, 2016). It has been present in the eXist code base ever since then. 

Does this jstack output help identify what's going on?

```text
Joes-MacBook-Pro-13-PAHO:exist joe$ jstack 17980
2018-10-03 16:46:01
Full thread dump Java HotSpot(TM) 64-Bit Server VM (11+28 mixed mode):

Threads class SMR info:
_java_thread_list=0x00007f93d6bcd310, length=16, elements={
0x00007f93d1808800, 0x00007f93d28d7800, 0x00007f93d2906800, 0x00007f93d181b800,
0x00007f93d2030000, 0x00007f93d202a800, 0x00007f93d183f000, 0x00007f93d202b800,
0x00007f93d1951800, 0x00007f93d2c5f000, 0x00007f93d139b800, 0x00007f93d2331800,
0x00007f93d2369800, 0x00007f93d1435000, 0x00007f93d1005000, 0x00007f93d2e8c800
}

"Reference Handler" #2 daemon prio=10 os_prio=31 cpu=0.80ms elapsed=154.46s tid=0x00007f93d1808800 nid=0x4903 waiting on condition  [0x0000700003322000]
   java.lang.Thread.State: RUNNABLE
	at java.lang.ref.Reference.waitForReferencePendingList(java.base@11/Native Method)
	at java.lang.ref.Reference.processPendingReferences(java.base@11/Reference.java:241)
	at java.lang.ref.Reference$ReferenceHandler.run(java.base@11/Reference.java:213)

"Finalizer" #3 daemon prio=8 os_prio=31 cpu=1.39ms elapsed=154.46s tid=0x00007f93d28d7800 nid=0x4703 in Object.wait()  [0x0000700003425000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(java.base@11/Native Method)
	- waiting on <no object reference available>
	at java.lang.ref.ReferenceQueue.remove(java.base@11/ReferenceQueue.java:155)
	- waiting to re-lock in wait() <0x00000007009e6ba0> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(java.base@11/ReferenceQueue.java:176)
	at java.lang.ref.Finalizer$FinalizerThread.run(java.base@11/Finalizer.java:170)

"Signal Dispatcher" #4 daemon prio=9 os_prio=31 cpu=0.23ms elapsed=154.45s tid=0x00007f93d2906800 nid=0x4203 runnable  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"C2 CompilerThread0" #5 daemon prio=9 os_prio=31 cpu=975.11ms elapsed=154.45s tid=0x00007f93d181b800 nid=0x4003 waiting on condition  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE
   No compile task

"C1 CompilerThread0" #8 daemon prio=9 os_prio=31 cpu=807.67ms elapsed=154.45s tid=0x00007f93d2030000 nid=0x3d03 waiting on condition  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE
   No compile task

"Sweeper thread" #9 daemon prio=9 os_prio=31 cpu=22.52ms elapsed=154.45s tid=0x00007f93d202a800 nid=0x5603 runnable  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Service Thread" #10 daemon prio=9 os_prio=31 cpu=0.04ms elapsed=154.43s tid=0x00007f93d183f000 nid=0x5703 runnable  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Common-Cleaner" #11 daemon prio=8 os_prio=31 cpu=1.35ms elapsed=154.42s tid=0x00007f93d202b800 nid=0x5903 in Object.wait()  [0x0000700003b3a000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(java.base@11/Native Method)
	- waiting on <no object reference available>
	at java.lang.ref.ReferenceQueue.remove(java.base@11/ReferenceQueue.java:155)
	- waiting to re-lock in wait() <0x00000007009e2c98> (a java.lang.ref.ReferenceQueue$Lock)
	at jdk.internal.ref.CleanerImpl.run(java.base@11/CleanerImpl.java:148)
	at java.lang.Thread.run(java.base@11/Thread.java:834)
	at jdk.internal.misc.InnocuousThread.run(java.base@11/InnocuousThread.java:134)

"AppKit Thread" #14 daemon prio=5 os_prio=31 cpu=490.07ms elapsed=153.67s tid=0x00007f93d1951800 nid=0x307 runnable  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"AWT-Shutdown" #15 prio=5 os_prio=31 cpu=1.30ms elapsed=153.64s tid=0x00007f93d2c5f000 nid=0x8103 in Object.wait()  [0x0000700004352000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(java.base@11/Native Method)
	- waiting on <0x0000000700e08e38> (a java.lang.Object)
	at java.lang.Object.wait(java.base@11/Object.java:328)
	at sun.awt.AWTAutoShutdown.run(java.desktop@11/AWTAutoShutdown.java:291)
	- waiting to re-lock in wait() <0x0000000700e08e38> (a java.lang.Object)
	at java.lang.Thread.run(java.base@11/Thread.java:834)

"Java2D Queue Flusher" #16 daemon prio=10 os_prio=31 cpu=106.65ms elapsed=153.57s tid=0x00007f93d139b800 nid=0x12a0b in Object.wait()  [0x0000700004767000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(java.base@11/Native Method)
	- waiting on <no object reference available>
	at sun.java2d.opengl.OGLRenderQueue$QueueFlusher.run(java.desktop@11/OGLRenderQueue.java:205)
	- waiting to re-lock in wait() <0x0000000700e1ac20> (a sun.java2d.opengl.OGLRenderQueue$QueueFlusher)
	at java.lang.Thread.run(java.base@11/Thread.java:834)

"Java2D Disposer" #17 daemon prio=10 os_prio=31 cpu=0.55ms elapsed=153.53s tid=0x00007f93d2331800 nid=0x5c07 in Object.wait()  [0x0000700003d40000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(java.base@11/Native Method)
	- waiting on <no object reference available>
	at java.lang.ref.ReferenceQueue.remove(java.base@11/ReferenceQueue.java:155)
	- waiting to re-lock in wait() <0x0000000700e23a58> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(java.base@11/ReferenceQueue.java:176)
	at sun.java2d.Disposer.run(java.desktop@11/Disposer.java:144)
	at java.lang.Thread.run(java.base@11/Thread.java:834)

"AWT-EventQueue-0" #18 prio=6 os_prio=31 cpu=304.45ms elapsed=152.85s tid=0x00007f93d2369800 nid=0xa007 waiting on condition  [0x0000700003c3d000]
   java.lang.Thread.State: WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@11/Native Method)
	- parking to wait for  <0x0000000700e2e5e8> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(java.base@11/LockSupport.java:194)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(java.base@11/AbstractQueuedSynchronizer.java:2081)
	at java.awt.EventQueue.getNextEvent(java.desktop@11/EventQueue.java:566)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(java.desktop@11/EventDispatchThread.java:190)
	at java.awt.EventDispatchThread.pumpEventsForFilter(java.desktop@11/EventDispatchThread.java:124)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(java.desktop@11/EventDispatchThread.java:113)
	at java.awt.EventDispatchThread.pumpEvents(java.desktop@11/EventDispatchThread.java:109)
	at java.awt.EventDispatchThread.pumpEvents(java.desktop@11/EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.run(java.desktop@11/EventDispatchThread.java:90)

"TimerQueue" #21 daemon prio=5 os_prio=31 cpu=2.64ms elapsed=152.15s tid=0x00007f93d1435000 nid=0xf047 waiting on condition  [0x0000700004bf6000]
   java.lang.Thread.State: WAITING (parking)
	at jdk.internal.misc.Unsafe.park(java.base@11/Native Method)
	- parking to wait for  <0x000000070d626788> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(java.base@11/LockSupport.java:194)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(java.base@11/AbstractQueuedSynchronizer.java:2081)
	at java.util.concurrent.DelayQueue.take(java.base@11/DelayQueue.java:217)
	at javax.swing.TimerQueue.run(java.desktop@11/TimerQueue.java:171)
	at java.lang.Thread.run(java.base@11/Thread.java:834)

"DestroyJavaVM" #23 prio=5 os_prio=31 cpu=1712.17ms elapsed=149.14s tid=0x00007f93d1005000 nid=0x2603 waiting on condition  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"Attach Listener" #24 daemon prio=9 os_prio=31 cpu=0.56ms elapsed=0.10s tid=0x00007f93d2e8c800 nid=0xf807 waiting on condition  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

"VM Thread" os_prio=31 cpu=24.83ms elapsed=154.46s tid=0x00007f93d201f000 nid=0x3203 runnable  

"GC Thread#0" os_prio=31 cpu=15.60ms elapsed=154.47s tid=0x00007f93d1006800 nid=0x2d03 runnable  

"GC Thread#1" os_prio=31 cpu=15.38ms elapsed=154.13s tid=0x00007f93d11e9000 nid=0x9f03 runnable  

"GC Thread#2" os_prio=31 cpu=15.64ms elapsed=154.13s tid=0x00007f93d5020000 nid=0x5e03 runnable  

"GC Thread#3" os_prio=31 cpu=15.63ms elapsed=154.13s tid=0x00007f93d3810000 nid=0x6003 runnable  

"GC Thread#4" os_prio=31 cpu=25.15ms elapsed=154.13s tid=0x00007f93d2ae9000 nid=0x6103 runnable  

"GC Thread#5" os_prio=31 cpu=15.86ms elapsed=154.13s tid=0x00007f93d2a4d000 nid=0x6303 runnable  

"G1 Main Marker" os_prio=31 cpu=0.52ms elapsed=154.47s tid=0x00007f93d1056800 nid=0x4f03 runnable  

"G1 Conc#0" os_prio=31 cpu=17.73ms elapsed=154.47s tid=0x00007f93d2800800 nid=0x4b03 runnable  

"G1 Conc#1" os_prio=31 cpu=16.28ms elapsed=153.53s tid=0x00007f93d2d16800 nid=0xea07 runnable  

"G1 Refine#0" os_prio=31 cpu=0.20ms elapsed=154.47s tid=0x00007f93d1804000 nid=0x2f03 runnable  

"G1 Young RemSet Sampling" os_prio=31 cpu=28.77ms elapsed=154.47s tid=0x00007f93d1057000 nid=0x3003 runnable  
"VM Periodic Task Thread" os_prio=31 cpu=104.54ms elapsed=154.43s tid=0x00007f93d1842000 nid=0xa503 waiting on condition  

JNI global refs: 107, weak refs: 35
```

**Comment by duncdrum** [comment]
I m going to go out on a limp, and group this with service wrapper issues.

**Comment by adamretter** [comment]
I now believe that this is a problem with the eXist-db Launcher. Whereby the `LauncherWrapper` starts as a process, shows the config dialog, and then starts the `Launcher` as a second process. The first process, the `LauncherWrapper` does not seem to exit!
